File: Anul 2/Semestrul 2/MPP/labs/Tema 5-proiect Server Csharp-Client Java/proiectSCcsharp/angajatNetworking/ClientRpcReflectionWorker.cs
using System;
using System.Collections.Generic;
using System.Net.Sockets;
using System.Runtime.Serialization.Formatters.Binary;
using System.Runtime.Serialization;
using System.Threading;
using proiectSCcsharp.services;
using proiectSCcsharp.domain;
using System.Reflection;
using log4net;
using System.IO;
using proiectSCcsharp.networking;

namespace proiectSCcsharp.networking
{
    public class ClientRpcReflectionWorker : IObserver //, Runnable
    {

        private static readonly ILog Logger = LogManager.GetLogger("ClientRpcReflectionWorker");

        private readonly IService service;
        private readonly TcpClient connection;
        private readonly NetworkStream stream;
        private readonly IFormatter formatter;
        private volatile bool connected;
        public ClientRpcReflectionWorker(IService service, TcpClient connection)
        {
            this.service = service;
            this.connection = connection;
            try
            {
                stream = connection.GetStream();
                formatter = new BinaryFormatter();
                connected = true;
            }
            catch (Exception e)
            {
                Console.WriteLine(e.StackTrace);
            }
        }

        public virtual void Run()
        {
            while (connected)
            {
                try
                {
                    var request = formatter.Deserialize(stream);
                    object response = HandleRequest((Request)request);
                    if (response != null)
                    {
                        SendResponse((Response)response);
                    }
                }
                catch (Exception e)
                {
                    Console.WriteLine(e.StackTrace);
                }

                try
                {
                    Thread.Sleep(1000);
                }
                catch (Exception e)
                {
                    Console.WriteLine(e.StackTrace);
                }
            }
            try
            {
                stream.Close();
                connection.Close();
            }
            catch (Exception e)
            {
                Console.WriteLine("Error " + e);
            }
        }

        private Response HandleRequest(Request request)
        {
            Response response = null;
            var handlerName = "Handle" + request.Type;
            Console.WriteLine("HandlerName " + handlerName);
            try
            {
                var method = this.GetType().GetMethod(handlerName, new[] { typeof(Request) });
                response = (Response)method.Invoke(this, new[] { request });
                Console.WriteLine("Method " + handlerName + " invoked");
            }
            catch (TargetInvocationException ex)
            {
                Console.WriteLine(ex.InnerException);
            }
            catch (AmbiguousMatchException ex)
            {
                Console.WriteLine(ex.Message);
            }
            catch (TargetParameterCountException ex)
            {
                Console.WriteLine(ex.Message);
            }
            catch (TargetException ex)
            {
                Console.WriteLine(ex.Message);
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message);
            }
            return response;
        }

        public Response HandleLOGIN(Request request)
        {
            Logger.InfoFormat("method entered: handleLOGIN with parameters {0}", request);
            Angajat angajat = (Angajat)request.Data;
            try
            {
                var foundAngajat = service.login(angajat, this);
                Logger.Info("Angajat logged in");
                return new Response.Builder().Type(ResponseType.OK).Data(foundAngajat).Build();
            }
            catch (MyException e)
            {
                connected = false;
                Logger.Error("Error in worker (solving method handleLOGIN): {0}", e);
                return new Response.Builder().Type(ResponseType.ERROR).Data(e.Message).Build();
            }
        }

        public Response HandleLOGOUT(Request request)
        {
            Logger.InfoFormat("method entered: handleLOGOUT with parameters {0}", request);
            Angajat angajat = (Angajat)request.Data;
            try
            {
                service.logout(angajat);
                connected = false;
                Logger.Info("User logged out");
                return new Response.Builder().Type(ResponseType.OK).Build(); ;
            }
            catch (MyException e)
            {
                Logger.Error("Error in worker (solving method handleLOGOUT): {0}", e);
                return new Response.Builder().Type(ResponseType.ERROR).Data(e.Message).Build();
            }
        }

        public Response HandleADD_BILET(Request request)
        {
            Logger.InfoFormat("method entered: handleADD_BILET with parameters {0}", request);
            Bilet bilet = (Bilet)request.Data;
            try
            {
                service.addBilet(bilet);
                Logger.Info("Bilet bought");
                return new Response.Builder().Type(ResponseType.OK).Data(bilet).Build();
            }
            catch (MyException e)
            {
                Logger.Error("Error in worker (solving method handleADD_BILET): {0}", e);
                return new Response.Builder().Type(ResponseType.ERROR).Data(e.Message).Build();
            }
        }
        public Response HandleGET_SPECTACOL(Request request)
        {

            Logger.InfoFormat("method entered: handleGET_SPECTACOL with parameters {0}", request);
            Console.WriteLine("method entered: handleGET_SPECTACOL with parameters");

            long id = (long)request.Data;
            try
            {
                var spectacol = service.getSpectacol(id);
                Logger.InfoFormat("Spectacol found {0}", spectacol);
                Console.WriteLine("Spectacol found " + spectacol);
                return new Response.Builder().Type(ResponseType.GET_SPECTACOL).Data(spectacol).Build();
            }
            catch (MyException e)
            {
                Logger.Error("Error in worker (solving method handleGET_SPECTACOL): {0}", e);
                Console.WriteLine("Error in worker (solving method handleGET_SPECTACOL): " + e);
                return new Response.Builder().Type(ResponseType.ERROR).Data(e.Message).Build();
            }
        }
        public Response HandleGET_ALL_SPECTACOLE(Request request)
        {
            Logger.InfoFormat("method entered: handleGET_ALL_SPECTACOLE with parameters {0}", request);
            Console.WriteLine("method entered: handleGET_ALL_SPECTACOLE with parameters");
            try
            {
                List<Spectacol> spectacole = service.getAllSpectacole();
                Logger.InfoFormat("Spectacole found {0}", spectacole);
                Console.WriteLine(@"Spectacole found {0}", spectacole);
                return new Response.Builder().Type(ResponseType.GET_ALL_SPECTACOLE).Data(spectacole).Build();
            }
            catch (MyException e)
            {
                Logger.Error("Error in worker (solving method handleGET_ALL_SPECTACOLE): {0}", e);
                Console.WriteLine(@"Error in worker (solving method handleGET_ALL_SPECTACOLE): {0}", e);
                return new Response.Builder().Type(ResponseType.ERROR).Data(e.Message).Build();
            }
        }

        private void SendResponse(Response response)
        {
            Console.WriteLine("sending response " + response);
            lock (stream)
            {
                formatter.Serialize(stream, response);
                stream.Flush();
            }
        }

        public void BiletAdded(Bilet bilet)
        {
            var resp = new Response.Builder().Type(ResponseType.SOLD_BILET).Data(bilet).Build();
            Logger.Info($"Bilet sold {bilet}");
            Console.WriteLine($@"Bilet sold {bilet}");
            try
            {
                SendResponse(resp);
                Logger.Info("Response sent");
                Console.WriteLine("Response sent");
            }
            catch (IOException e)
            {
                Console.WriteLine(e);
                Logger.Error($"Error in worker (sending response): {e}");
                Console.WriteLine(@"Error in worker (sending response): {0}", e);
            }
        }
    }

}

File: Anul 2/Semestrul 2/MPP/labs/Tema 6-proiect Java ORM/proiectServerClientJava/Networking/src/main/java/proiectServerClientJava/network/rpcprotocol/ClientRpcReflectionWorker.java
package proiectServerClientJava.network.rpcprotocol;

import proiectServerClientJava.domain.*;
import proiectServerClientJava.domain.validators.ValidationException;
import proiectServerClientJava.service.IObserver;
import proiectServerClientJava.service.IService;
import proiectServerClientJava.service.MyException;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.io.IOException;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.net.Socket;
import java.util.List;


public class ClientRpcReflectionWorker implements Runnable, IObserver {
    private final IService service;
    private final Socket connection;

    private static final Logger logger= LogManager.getLogger();

    private ObjectInputStream input;
    private ObjectOutputStream output;
    private volatile boolean connected;
    public ClientRpcReflectionWorker(IService service, Socket connection) {
        logger.info("Creating worker");
        this.service = service;
        this.connection = connection;
        try{
            output=new ObjectOutputStream(connection.getOutputStream());
            output.flush();
            input=new ObjectInputStream(connection.getInputStream());
            connected=true;
            logger.info("Worker created");
        } catch (IOException e) {
            logger.error(e);
        }
    }

    public void run() {
        while(connected){
            try {
                Object request=input.readObject();
                Response response=handleRequest((Request)request);
                if (response!=null){
                    sendResponse(response);
                }
            } catch (IOException | ClassNotFoundException e) {
                logger.error("Error in worker (reading): "+e);
            }
            try {
                Thread.sleep(1000);
            } catch (InterruptedException e) {
                logger.error("Error in worker (sleeping): "+e);
            }
        }
        try {
            input.close();
            output.close();
            connection.close();
        } catch (IOException e) {
            logger.error("Error in worker (closing connection): "+e);
        }
    }

    private static final Response okResponse=new Response.Builder().type(ResponseType.OK).build();
    private Response handleRequest(Request request){
        Response response=null;
        String handlerName="handle"+(request).type();
        logger.traceEntry("method entered: "+handlerName+" with parameters "+request);
        try {
            Method method=this.getClass().getDeclaredMethod(handlerName, Request.class);
            response=(Response)method.invoke(this,request);
            logger.info("Method invoked: "+handlerName+" with response "+response);
        } catch (NoSuchMethodException | InvocationTargetException | IllegalAccessException e) {
            e.printStackTrace();
            logger.error("Error in worker (invoking method handleRequest): "+e);
        }
        return response;
    }

    private Response handleADD_BILET(Request request){
        logger.traceEntry("method entered: handleADD_BILET with parameters "+request);
        Bilet bilet=(Bilet) request.data();
        try {
            service.addBilet(bilet);
            logger.info("Bilet bought");
            return new Response.Builder().type(ResponseType.OK).data(bilet).build();
        } catch (MyException e) {
            logger.error("Error in worker (solving method handleADD_BILET): "+e);
            return new Response.Builder().type(ResponseType.ERROR).data(e.getMessage()).build();
        } catch (ValidationException e) {
            throw new RuntimeException(e);
        }
    }

    private Response handleGET_SPECTACOL(Request request) {
        logger.traceEntry("method entered: handleGET_SPECTACOL with parameters " + request);
        Long id = (Long) request.data();
        try {
            Spectacol spectacol = service.getSpectacol(id);
            logger.info("Spectacol found " + spectacol);
            return new Response.Builder().type(ResponseType.GET_SPECTACOL).data(spectacol).build();
        } catch (MyException e) {
            logger.error("Error in worker (solving method handleGET_SPECTACOL): " + e);
            return new Response.Builder().type(ResponseType.ERROR).data(e.getMessage()).build();
        }
    }

    private Response handleGET_ALL_SPECTACOLE(Request request){
        logger.traceEntry("method entered: handleGET_ALL_SPECTACOLE with parameters "+request);
        try {
            List<Spectacol> spectacole= (List<Spectacol>) service.getAllSpectacole();
            logger.info("Spectacole found "+spectacole);
            return new Response.Builder().type(ResponseType.GET_ALL_SPECTACOLE).data(spectacole).build();
        } catch (MyException e) {
            logger.error("Error in worker (solving method handleGET_ALL_SPECTACOLE): "+e);
            return new Response.Builder().type(ResponseType.ERROR).data(e.getMessage()).build();
        }
    }
    private Response handleLOGIN(Request request){
        logger.traceEntry("method entered: handleLOGIN with parameters "+request);
        Angajat employee = (Angajat) request.data();
        try {
            Angajat foundEmployee = service.login(employee, this);
            logger.info("Employee logged in");
            return new Response.Builder().type(ResponseType.OK).data(foundEmployee).build();
        } catch (MyException e) {
            connected=false;
            logger.error("Error in worker (solving method handleLOGIN): "+e);
            return new Response.Builder().type(ResponseType.ERROR).data(e.getMessage()).build();
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    private Response handleLOGOUT(Request request){
        logger.traceEntry("method entered: handleLOGOUT with parameters "+request);
        Angajat employee=(Angajat) request.data();
        try {
            service.logout(employee, this);
            connected=false;
            logger.info("User logged out");
            return okResponse;
        } catch (MyException e) {
            logger.error("Error in worker (solving method handleLOGOUT): "+e);
            return new Response.Builder().type(ResponseType.ERROR).data(e.getMessage()).build();
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }
    private void sendResponse(Response response) throws IOException{
        logger.traceEntry("method entered: sendResponse with parameters "+response);
        output.writeObject(response);
        output.flush();
        logger.info("Response sent");
    }

    @Override
    public void biletAdded(Bilet bilet) throws MyException {
        Response resp=new Response.Builder().type(ResponseType.SOLD_BILET).data(bilet).build();
        logger.info("Bilet sold "+bilet);
        try {
            sendResponse(resp);
            logger.info("Response sent");
        } catch (IOException e) {
            e.printStackTrace();
            logger.error("Error in worker (sending response): "+e);
        }
    }
}